RFC-0027-PyTorch-Distributed-Tensor.md

```diff
@@ -0,0 +1,155 @@
+
+## Value Propsition
+
+The primary value of DistributedTensor includes:
+-   Offers a uniform way to save/load state dict during checkpointing, even when there’re complex data distribution strategies
+-   DistributedTensor could natively offer Tensor Parallelism solution in eager mode, just like our current ShardedTensor solution. Moreover, it gives additional flexibility for advanced users who want to mix sharding and replication.
+-   DistributedTensor could be used as a basic building block of a compiler based solution to do distributed training    
+-   DistributedTensor could act as a SPMD programming model entry point for ML System Engineers, providing good UX to mix up different types of parallelism.
+
+## PyTorch Distributed Tensor
+
+### DistributedTensor API
+
+We offer both lower level DistributedTensor API and module level API to create nn.Module with “distributed” parameters.
+
+#### Basic DistributedTensor API Examples
+
+Here are some basic DistributedTensor level API examples that showcase: 
+1.  How to construct a DistributedTensor directly, to represent different types of sharding, replication, sharding + replication
+2. How to create DistributedTensor from a local torch.Tensor
+3. How to “reshard” an existing DistributedTensor to a different DistributedTensor with different placement strategy, or world size.
+
+```python
+import torch  
+import torch.distributed as distributed  
+from torch.distributed import DTensor, DeviceMesh, Shard, Replicate, distribute_module  
+  
+# construct a device mesh with available devices (multi-host or single host)  
+device_mesh = DeviceMesh(device_type="cuda", [0, 1, 2, 3])  
+# if we want to do row-wise sharding  
+rowwise_placement=[Shard(0)]  
+# if we want to do col-wise sharding  
+colwise_placement=[Shard(1)]  
+# distributed tensor returned will be sharded across the dimension specified in placements  
+distributed.empty((8, 12), device_mesh=device_mesh, placements=rowwise_placement)  
+  
+# if we want to do replication across a certain device list  
+replica_placement = [Replicate()]  
+# distributed tensor will be replicated to all four GPUs.  
+distributed.empty((8, 12), device_mesh=device_mesh, placements=replica_placement)  
+  
+# if we want to distributed a tensor with both replication and sharding  
+device_mesh = DeviceMesh(device_type="cuda", [[0, 1], [2, 3]])  
+# replicate across the first dimension of device mesh, then sharding on the second dimension of device mesh  
+spec=[Replicate(), Shard(0)]  
+distributed.empty((8, 8), device_mesh=device_mesh, placements=spec)  
+  
+# create a DistributedTensor that shards on dim 0, from a local torch.Tensor  
+local_tensor = torch.randn((8, 8), requires_grad=True)  
+rowwise_tensor = DTensor.from_local(local_tensor, device_mesh, rowwise_placement)  
```

Comment:
Can we add a comment about how `from_local` is different from calling DTensor(torch.empty())?